pageserver/src/tenant.rs

```diff
@@ -1639,64 +1644,88 @@ impl Tenant {
                     info!("Tenant is already in Stopping state, skipping activation");
                 }
                 TenantState::Loading | TenantState::Attaching => {
-                    *current_state = TenantState::Active;
-
+                    *current_state = TenantState::Activating;
                     debug!(tenant_id = %self.tenant_id, "Activating tenant");
+                    activating = true;
+                    // Continue outside the closure. We need to grab timelines.lock()
+                    // and we plan to turn it into a tokio::sync::Mutex in a future patch.
+                }
+            }
+        });
+        if let Err(e) = result {
+            assert!(!activating, "transition into Activating is infallible");
+            return Err(e);
+        }
 
-                    let timelines_accessor = self.timelines.lock().unwrap();
-                    let not_broken_timelines = timelines_accessor
-                        .values()
-                        .filter(|timeline| timeline.current_state() != TimelineState::Broken);
+        if activating {
+            let timelines_accessor = self.timelines.lock().unwrap();
+            let not_broken_timelines = timelines_accessor
+                .values()
+                .filter(|timeline| timeline.current_state() != TimelineState::Broken);
 
-                    // Spawn gc and compaction loops. The loops will shut themselves
-                    // down when they notice that the tenant is inactive.
-                    tasks::start_background_loops(self.tenant_id);
+            // Spawn gc and compaction loops. The loops will shut themselves
+            // down when they notice that the tenant is inactive.
+            tasks::start_background_loops(self.tenant_id);
 
-                    let mut activated_timelines = 0;
+            let mut activated_timelines = 0;
 
-                    for timeline in not_broken_timelines {
-                        timeline.activate(broker_client.clone(), ctx);
-                        activated_timelines += 1;
-                    }
+            for timeline in not_broken_timelines {
+                timeline.activate(broker_client.clone(), ctx);
+                activated_timelines += 1;
+            }
 
-                    let elapsed = self.loading_started_at.elapsed();
-                    let total_timelines = timelines_accessor.len();
+            self.state.send_modify(move |current_state| {
+                assert!(
+                    *current_state == TenantState::Activating,
+                    "set_stopping and set_broken wait for us to leave Activating state",
+                );
```

Comment:
`assert_eq!` will in all cases give out a better assertion failure, so better use it here, add PartialEq if needed ... Even though it does feel icky if missing for the Broken state, let's just tolerate that.

```suggestion
                assert_eq!(
                    *current_state,
                    TenantState::Activating,
                    "set_stopping and set_broken wait for us to leave Activating state",
                );
```